File: DataRanker_Backend/services/ranker.py
# services/ranker.py — Stage 3: Score and rank companies by KPI template

import numpy as np
import pandas as pd
from openpyxl.styles import PatternFill

from core.config import KPI_HEADER_ROW, KPI_SHEET_NAME, METRIC_SCORE_COLOR, RANK_COLOR


# ── Helpers ───────────────────────────────────────────────────────────────────

def _make_fill(hex_color: str) -> PatternFill:
    return PatternFill(start_color=hex_color, end_color=hex_color, fill_type="solid")


def _score_kpi(
    subset: pd.DataFrame,
    kpi: str,
    weight: float,
    direction: str,
) -> tuple[pd.DataFrame, str]:
    """
    Compute rank, percentile slab, and weighted metric score for one KPI.

    Direction semantics (from the KPI Library "Higher/Lower Better" column):
      - "Higher" → a larger value is better → rank 1 goes to the largest value
                   → pandas rank with ascending=False
      - "Lower"  → a smaller value is better → rank 1 goes to the smallest value
                   → pandas rank with ascending=True

    The _Rank column uses the same convention so rank 1 always means "best".
    The percentile slab is derived the same way so that the best company always
    lands in the top slab (100) and the worst in the bottom slab (10), regardless
    of direction.

    Returns the mutated subset and the name of the new metric-score column.
    """
    higher_is_better: bool = direction.strip().lower() == "higher"

    # Rank 1 = best company for this KPI
    #   "Higher" → largest value is best → ascending=False so rank 1 → largest
    #   "Lower"  → smallest value is best → ascending=True  so rank 1 → smallest
    subset[f"{kpi}_Rank"] = subset[kpi].rank(ascending=not higher_is_better)

    # Percentile slab: best company → 100th percentile, worst → 10th percentile
    #   pandas rank(pct=True, ascending=True)  → largest value gets pct≈1.0
    #   pandas rank(pct=True, ascending=False) → smallest value gets pct≈1.0
    #   "Higher" → want largest  to score highest → ascending=True
    #   "Lower"  → want smallest to score highest → ascending=False
    raw_pct = subset[kpi].rank(pct=True, ascending=higher_is_better) * 100
    subset[f"{kpi}_Percentile_Slab"] = np.ceil(raw_pct / 10) * 10

    score_col = f"{kpi}_Metric_Score"
    subset[score_col] = subset[f"{kpi}_Percentile_Slab"] * weight

    return subset, score_col


def _apply_cell_fills(sheet, columns: list[str], n_rows: int) -> None:
    """Colour metric-score and company-rank columns in the given sheet."""
    metric_fill = _make_fill(METRIC_SCORE_COLOR)
    rank_fill = _make_fill(RANK_COLOR)

    for col_idx, col_name in enumerate(columns, 1):
        if "_Metric_Score" in col_name:
            fill = metric_fill
        elif col_name == "Company_Rank":
            fill = rank_fill
        else:
            continue

        for row_idx in range(2, n_rows + 2):
            sheet.cell(row=row_idx, column=col_idx).fill = fill


# ── Public entry point ────────────────────────────────────────────────────────

def run_ranking(mapped_path: str, kpi_lib_path: str, out_path: str) -> None:
    """
    For every KPI template found in the mapped CSV, score and rank companies
    using the KPI library weights, then write each template to its own sheet.

    Args:
        mapped_path:  Path to the Stage-2 mapped CSV.
        kpi_lib_path: Path to the KPI library .xlsx workbook.
        out_path:     Path where the final ranked .xlsx will be written.
    """
    df = pd.read_csv(mapped_path)
    kpi_lib = pd.read_excel(kpi_lib_path, sheet_name=KPI_SHEET_NAME, header=KPI_HEADER_ROW)
    kpi_lib.columns = kpi_lib.columns.str.strip()

    with pd.ExcelWriter(out_path, engine="openpyxl") as writer:
        for template in df["KPI_Template"].dropna().unique():
            _write_template_sheet(writer, df, kpi_lib, template)


def _write_template_sheet(
    writer: pd.ExcelWriter,
    df: pd.DataFrame,
    kpi_lib: pd.DataFrame,
    template: str,
) -> None:
    """Score, rank, and write one template's companies to its own Excel sheet."""
    subset = df[df["KPI_Template"] == template].copy()
    metrics = kpi_lib[kpi_lib["Template"] == template]

    score_cols: list[str] = []
    for _, row in metrics.iterrows():
        kpi = row["KPI"]
        weight = row["Weight %"] / 100
        direction = row["Higher/Lower Better"]

        # FIX 3: A column may exist but be entirely NaN because formatter.py
        # pre-created it as a placeholder (EXTRA_COLUMNS). Scoring an all-NaN
        # series produces all-NaN _Rank / _Percentile_Slab / _Metric_Score cols.
        # Skip any KPI whose values are all missing.
        if kpi in subset.columns and subset[kpi].notna().any():
            subset, score_col = _score_kpi(subset, kpi, weight, direction)
            score_cols.append(score_col)

    subset["Total_Final_Score"] = subset[score_cols].sum(axis=1)
    subset["Company_Rank"] = subset["Total_Final_Score"].rank(
        ascending=False, method="min"
    )

    sheet_name = str(template)[:31]
    subset.to_excel(writer, sheet_name=sheet_name, index=False)

    _apply_cell_fills(writer.sheets[sheet_name], list(subset.columns), len(subset))

File: server/services/ranker.js
// services/ranker.js — Stage 3: score and rank companies by KPI template.
//
// Port of ranker.py.

const { rank } = require("../lib/rank");
const { readXlsxWithHeader, writeXlsx } = require("../lib/io");
const {
  KPI_HEADER_ROW,
  KPI_SHEET_NAME,
  METRIC_SCORE_COLOR,
  RANK_COLOR,
} = require("../core/config");

function isNum(v) {
  return typeof v === "number" && Number.isFinite(v);
}

// Score one KPI on the subset. Mutates subset rows (adds _Rank,
// _Percentile_Slab, _Metric_Score) and returns the metric-score column name.
function scoreKpi(subset, columns, kpi, weight, direction) {
  const higherIsBetter = String(direction).trim().toLowerCase() === "higher";

  const values = subset.map((r) => {
    const v = r[kpi];
    return isNum(v) ? v : null;
  });

  // Rank 1 = best: "Higher" → ascending=false; "Lower" → ascending=true.
  const rankVals = rank(values, { ascending: !higherIsBetter });

  // Percentile slab: pct=true, ascending=higherIsBetter.
  const pctVals = rank(values, { ascending: higherIsBetter, pct: true });

  const rankCol = `${kpi}_Rank`;
  const slabCol = `${kpi}_Percentile_Slab`;
  const scoreCol = `${kpi}_Metric_Score`;

  for (let i = 0; i < subset.length; i++) {
    subset[i][rankCol] = rankVals[i]; // null stays null

    let slab = null;
    if (pctVals[i] !== null) {
      const rawPct = pctVals[i] * 100;
      slab = Math.ceil(rawPct / 10) * 10;
    }
    subset[i][slabCol] = slab;

    subset[i][scoreCol] = slab === null ? null : slab * weight;
  }

  for (const c of [rankCol, slabCol, scoreCol]) {
    if (!columns.includes(c)) columns.push(c);
  }

  return scoreCol;
}

// mapped: { columns, rows } from Stage 2.
async function runRanking(mapped, kpiBuffer) {
  // Read KPI library, header on 0-indexed row KPI_HEADER_ROW; strip col names.
  const kpiTable = readXlsxWithHeader(kpiBuffer, KPI_SHEET_NAME, KPI_HEADER_ROW);
  const kpiStripped = kpiTable.columns.map((c) => String(c).trim());
  const kpiRows = kpiTable.rows.map((r) => {
    const obj = {};
    for (let i = 0; i < kpiTable.columns.length; i++) {
      obj[kpiStripped[i]] = r[kpiTable.columns[i]];
    }
    return obj;
  });

  // Distinct non-null KPI_Template values in first-appearance order.
  const seen = new Set();
  const templates = [];
  for (const row of mapped.rows) {
    const t = row["KPI_Template"];
    if (t === null || t === undefined || (typeof t === "number" && Number.isNaN(t))) continue;
    if (!seen.has(t)) {
      seen.add(t);
      templates.push(t);
    }
  }

  const sheets = [];
  for (const template of templates) {
    sheets.push(buildTemplateSheet(mapped, kpiRows, template));
  }

  return writeXlsx(sheets);
}

function buildTemplateSheet(mapped, kpiRows, template) {
  // subset = deep-enough copy of rows where KPI_Template === template.
  const subset = mapped.rows
    .filter((r) => r["KPI_Template"] === template)
    .map((r) => Object.assign({}, r));
  const columns = mapped.columns.slice();

  const metrics = kpiRows.filter((m) => m["Template"] === template);

  const scoreCols = [];
  for (const row of metrics) {
    const kpi = row["KPI"];
    const weight = row["Weight %"] / 100;
    const direction = row["Higher/Lower Better"];

    // Skip if kpi not a column in subset, or all subset values null/NaN.
    if (!columns.includes(kpi)) continue;
    const anyValid = subset.some((r) => isNum(r[kpi]));
    if (!anyValid) continue;

    const scoreCol = scoreKpi(subset, columns, kpi, weight, direction);
    scoreCols.push(scoreCol);
  }

  // Total_Final_Score = row-wise sum of *_Metric_Score, skipna; all-null → 0.
  for (const r of subset) {
    let sum = 0;
    for (const sc of scoreCols) {
      const v = r[sc];
      if (isNum(v)) sum += v;
    }
    r["Total_Final_Score"] = sum;
  }
  if (!columns.includes("Total_Final_Score")) columns.push("Total_Final_Score");

  // Company_Rank = rank(Total_Final_Score, ascending=false, method="min").
  const totals = subset.map((r) => r["Total_Final_Score"]);
  const companyRanks = rank(totals, { ascending: false, method: "min" });
  for (let i = 0; i < subset.length; i++) {
    subset[i]["Company_Rank"] = companyRanks[i];
  }
  if (!columns.includes("Company_Rank")) columns.push("Company_Rank");

  const sheetName = String(template).slice(0, 31);

  // Cell fills: columns containing "_Metric_Score" → METRIC_SCORE_COLOR;
  // column exactly "Company_Rank" → RANK_COLOR. (Data rows only.)
  const fills = [];
  for (let ci = 0; ci < columns.length; ci++) {
    const name = columns[ci];
    if (name.includes("_Metric_Score")) {
      fills.push({ colIndex: ci, argb: "FF" + METRIC_SCORE_COLOR });
    } else if (name === "Company_Rank") {
      fills.push({ colIndex: ci, argb: "FF" + RANK_COLOR });
    }
  }

  return { name: sheetName, columns, rows: subset, fills };
}

module.exports = { runRanking };
